LIN-736: Merge dvc_dag_PythonOperator.jinja into task_function.jinja

andycui97 opened the pull request:
# Description

Merges repeated logic in dvc_dag_PythonOperator.jinja into task_function.jinja. 

Changes: 
* Rename localpickle to tmpdirpickle to better match the behavior of the serialization flavor. 
    * Add CWD pickle flavor
* Remove stages logic in DVC pipeline writer to directly use TaskDefinitions where possible and avoid intermediate task representation specific to DVC. 
    * Change lineapy/plugins/jinja_templates/dvc/dvc_dag_StagePerArtifact.jinja to use Task Definitions as well. 

* Change [dvc_dag_PythonOperator.jinja](https://github.com/LineaLabs/lineapy/pull/872/files#diff-6f790f9c60d75b1d9060cb5c3081a053feb39f0115def96b36bab0cb18659642) to use TASK_CODE block generated from the task_function template. 

Fixes: LIN-736

## Type of change

Please delete options that are not relevant.

- [X] Refactor (non-breaking change which should not change functionality)

# How Has This Been Tested?

- Ensure snapshot tests for DVC do not change. DVC output has been prettified so the snapshot diff has a lot of changes which have been verified to be only spacing and formatting changes.

aayan636 reviewed (commented):
My. main query here is how does renaming `LocalPickle` to `TmpDirPickle` help in the PR. Most probably I'm missing some context here so a sync up would be helpful!

andycui97 commented on lineapy/plugins/task.py:
```diff
@@ -179,9 +179,11 @@ class TaskSerializer(Enum):
+    TmpDirPickle = 1
     # Write to a pickle directory that can be parametrized
     ParametrizedPickle = 2
+    # Write pickle to same directory as pipeline files
```
current working directory

andycui97 commented on lineapy/plugins/airflow_pipeline_writer.py:
```diff
@@ -164,11 +164,11 @@ def task_iris_pred():
         )
 
         # Add setup and teardown if local pickle serializer is selected
```
Not a nit, good catch I updated the comments to get setup and teardown using this serialization as well.

andycui97 commented:
> My. main query here is how does renaming `LocalPickle` to `TmpDirPickle` help in the PR. Most probably I'm missing some context here so a sync up would be helpful!

I renamed `LocalPickle` because it actually describes the newly added `CWD pickle` flavor. LocalPickle suggests we are using Pickle to a local directory but that doesnt distinguish between using local tmp vs local working directory. I renamed them both so that there is a distinction.

andycui97 commented on tests/unit/plugins/framework_specific/dvc/__snapshots__/test_writer_dvc.ambr:
```diff
@@ -676,33 +676,46 @@
 ---
```
Yes I ran it on everything. This is because I also prettified the DVC output here: https://github.com/LineaLabs/lineapy/blob/7e17ddbd072ce78cbf296a175799cc0b8ff69435/lineapy/plugins/dvc_pipeline_writer.py#L175 

The actual logic of the files havent changed (please verify as part of review) just the spacing and formatting.

aayan636 reviewed (commented):
Just one minor comment on a new line in the dvc snapshot, rest looks good to me. Since the builds are passing and apart from one the snapshots hav not changed, seems like the refactor is sound so should be good to go after the snapshot change is looked into.

andycui97 commented on tests/unit/plugins/framework_specific/dvc/__snapshots__/test_writer_dvc.ambr:
```diff
@@ -1147,33 +1202,46 @@
   
   def task_b0(b0):
+  
+      b0 = int(b0)
```
It comes from the typing_block in the task template. 

https://github.com/LineaLabs/lineapy/blob/93c305581a3018161341b7b936c9d1c8e5be89d4/lineapy/plugins/jinja_templates/task/task_function.jinja#L8 

which is provided here: 

https://github.com/LineaLabs/lineapy/pull/872/files#diff-985b13e42b23663cc753311b147b2a2e92fb07c8354417f8ee5acc1af248572eR154 

where the old template with the stages did not have this at all. I actually think providing this line is necessary for correctness. 

Great catch.
